### DISCUSSION TITLE
Running two instances of a container with persistent data & mapped ports

### DISCUSSION THREAD

**Discussion by adamshand** [caprover/caprover#2007] [Q&A]
I have a two node CapRover instance.  I'd like to run an instance of AdGuardHome on each node for redundancy.  AGH requires a persistent volume and also the DNS and DHCP ports (53, 67, 68 & 853) mapped through to the container.

I'm struggling to find a way to do this.  

At first, I thought I could just add a second replica and that would create another instance with it's own volume which I could manually configure.  But the replicas share the same volume.

Next, I thought I'd be able to create a second instance of AGH and then manually bind it to the worker node. That doesn't work because there can only be one container per-swarm with a given mapped port.

The internet suggested I could change to publishing posts in host mode (instead of mesh) and that should allow me to have port 53 mapped to a container on each swarm node.  So I removed all the mapped ports from the primay AGH instance, created a second one (removed it's maps ports) and moved it to my worker node.  Then tried to add host mode port mapping like:

```
docker service update \
  --publish-add mode=host,published=53,target=53,protocol=udp \
  --publish-add mode=host,published=53,target=53,protocol=tcp \
  --publish-add mode=host,published=67,target=67,protocol=udp \
  --publish-add mode=host,published=68,target=68,protocol=udp \
  --publish-add mode=host,published=853,target=853,protocol=tcp \
  srv-captain--adguard
  ```
 
  And swarm seemed to choke on that.
  
  Am I missing something obvious, or am I just trying to do something that Swarm is ill-suited to?
  
  Thanks for any pointers.

**Comment by githubsaturn** [comment]
You shouldn't use CapRover's replicas in this case. You should create two separate apps and manually change their NodeID to the respective nodes.

**Reply by adamshand** [reply]
Thanks, I figured that out.  

But CapRover won't let me create a second instance if it tries to map a port (eg. 53/UDP) which an existing container already has mapped.

**Reply by githubsaturn** [reply]
you should use service override and set mode to host, let me know if that doesn't work

**Reply by adamshand** [reply]
So I can change the mode to host via a service override and that seems to work properly:

```
❯ docker service inspect srv-captain--adguard2 | jq '.[].Endpoint.Spec.Ports[]'
{
  "Protocol": "udp",
  "TargetPort": 53,
  "PublishedPort": 53,
  "PublishMode": "host"
}, 
{ ... }
``` 

~~But when I change the application node id to the worker node, CapRover locks up.  The web interface stops responding.  It takes a long time (5 minutes ish?) but it eventually starts responding again.~~

However, after this CapRover shows that there is one instance running, but docker claims there are 0.

```
qgkil75wrxbn   srv-captain--adguard2           replicated   0/1        adguard/adguardhome:v0.107.44
```

![image](https://github.com/caprover/caprover/assets/130651/93683fa5-18dd-4857-a2f7-320dcf218135)

If I reset the CapRover instance count to zero and then back to 1, I get this in the logs:

```
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:50.601 pm    Updating app started: adguard2
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:50.613 pm    Ensure service inited and Updated for: adguard2
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:50.615 pm    Check if service is running: srv-captain--adguard2
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:50.618 pm    Service is already running: srv-captain--adguard2
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:50.619 pm    Updating service srv-captain--adguard2 with image adguard/adguardhome:v0.107.44
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.631 pm    Updating Load Balancer - ServiceManager
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.632 pm    Locking NGINX configuration reloading...
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.657 pm    SUCCESS: UNLocking NGINX configuration reloading...
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.657 pm    sendReloadSignal...
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.666 pm    Kill HUP Container: 0fb58b6244e18ffe52e824fbf6e804c3a8cb9e562f4012ca1fe17190497b022f
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.696 pm    Updating Load Balancer - ServiceManager
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.696 pm    Locking NGINX configuration reloading...
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.702 pm    SUCCESS: UNLocking NGINX configuration reloading...
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.702 pm    sendReloadSignal...
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.704 pm    Kill HUP Container: 0fb58b6244e18ffe52e824fbf6e804c3a8cb9e562f4012ca1fe17190497b022f
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:52.722 pm    AppName is updated: adguard2
captain-captain.1.zb82i7ln09kc@ponga    | POST /api/v2/user/apps/appDefinitions/update 200 2123.710 ms - 69
captain-captain.1.zb82i7ln09kc@ponga    | GET /api/v2/user/apps/appDefinitions 200 1.396 ms - 11340
captain-captain.1.zb82i7ln09kc@ponga    | February 16th 2024, 10:07:55.631 pm    Pruning containers...
```

But while CapRover seems to think everything is working, swarm doesn't:

```
❯ docker service ls | grep adguard2
qgkil75wrxbn   srv-captain--adguard2           replicated   0/1        adguard/adguardhome:v0.107.44
```

**Reply by adamshand** [reply]
Moving AdGuardHome's node id back to the master node and everything starts working as expected again.

**Comment by adamshand** [comment]
If anyone wants to share ports in host mode via a service override, this is a working YAML override for AdGuardHome.

```yaml
EndpointSpec:
  Ports:
    - Protocol: udp
      TargetPort: 53
      PublishedPort: 53
      PublishMode: host
    - Protocol: tcp
      TargetPort: 53
      PublishedPort: 53
      PublishMode: host
    - Protocol: tcp
      TargetPort: 80
      PublishedPort: 8080
      PublishMode: host
    - Protocol: tcp
      TargetPort: 853
      PublishedPort: 853
      PublishMode: host
    - Protocol: udp
      TargetPort: 67
      PublishedPort: 67
      PublishMode: host
    - Protocol: tcp
      TargetPort: 68
      PublishedPort: 68
      PublishMode: host
```
